## Calculating Dominant Phenotype Probabilities in Matings using Complementary Probability

### Domain: Computational Biology | Population Genetics | Mendelian Inheritance
### Benchmark Dataset Source: ROSALIND - Mendel's First Law

### 1. Project Abstract & Biological Context

- Mendel’s First Law (the Law of Segregation) establishes that diploid organisms possess two alleles for every gene, which segregate during gamete  -- formation so that each gamete carries only a single allele. 

- Mendel’s First Law (the Law of Segregation) establishes that diploid organisms possess two alleles for every gene, which segregate during gamete formation so that each gamete carries only a single allele. In population genetics, evaluating the likelihood of inheriting dominant traits across random matings provides a baseline for understanding trait transmission and allele persistence.

- In population genetics, evaluating the likelihood of inheriting dominant traits across random matings provides a baseline for understanding trait transmission and allele persistence.

### Experimental Framework

Consider a diploid population consisting of three distinct genotype classes:
- k Homozygous Dominant (AA): Always contributes a dominant allele (A).
- m Heterozygous (Aa): Has a 50% chance of passing A and a 50% chance of passing a.
- n Homozygous Recessive (aa): Always contributes a recessive allele (a).

### Mathematical Logic (Complementary Probability)

Instead of summing all parent combinations that produce dominant offspring, it is computationally simpler to use complementary probability:
                        P(Dominant Phenotype) = 1 - P(Recessive Offspring)

An offspring can display the recessive phenotype (aa) only if both parents pass a recessive allele (a). The only parental pairings capable of producing an aa child are:
- aa x aa:     Both parents homozygous recessive (100% chance of aa offspring).
- Aa x Aa:     Both parents heterozygous (25% chance of aa offspring).
- Aa x aa:     (or aa x Aa): One heterozygous and one homozygous recessive parent (50% chance of aa offspring).

Given total population size T = k + m + n, the total number of ordered parental selections without replacement is T(T - 1).


### 2. Mathematical Framework & Equation Derivation

#### Weighted Recessive Offspring Count

Summing the weighted outcomes across all recessive-producing parental pairs gives:

- Recessive Outcomes = n(n - 1) x 1.0 + m(m - 1) x 0.25 + (2 x m x n x 0.5)
                        aa x aa             Aa x Aa           Aa x aa/Aa


- Dividing by total ordered pairings T(T - 1) gives the probability of a recessive outcome:
P(Recessive) = n(n - 1) + 0.25 x m(m - 1) + m x n
               -----------------------------------
                          T(T - 1)


- Final Closed-Form Equation:
                    _                                   _
                   | n(n - 1) + 0.25 x m(m - 1) + m x n  |
P(Dominant) = 1 -  | ----------------------------------- |
                   |              T(T - 1)               |  
                   |_                                   _|

## Python Implementation Strategies

Below are two software implementations for predicting dominant phenotype probabilities:

- 1) Pure Python Function: An explicit, parameter-driven function using exact population counts.
- 2) Dictionary & Vectorized Parsing: A structured approach reading from file input into key-value genotype mappings.


## The Goal:

Consider a population defined by genotype parameters k = 29, m = 23, and n = 26, 
where k, m, and n represent homozygous dominant (AA), heterozygous (Aa), and homozygous recessive (aa) individuals, respectively. 

- a) Part A: Determine the total population size T and the total number of ordered parent pairs T(T - 1).
- b) Part B: Calculate the total weighted number of outcomes that yield a homozygous recessive (aa) offspring.
- c) Part C: Use complementary probability to find P({Dominant Phenotype}). Express your final answer rounded to 5 decimal places.

Goal: Given a population containing a mix of homozygous dominant (k), heterozygous (m), and homozygous recessive (n) individuals, calculate the probability that two randomly chosen individuals will produce an offspring that expresses the dominant phenotype.

### Method 1: Pure Python Parameter-Driven Function

In [101]:
def get_dominant_probability(k: int, m: int, n: int) -> float:
    # Total population size
    total_pop = k + m + n
    total_pairs = total_pop * (total_pop - 1)     # Total possible ordered pairings

    # Weighted counts of pairings producing recessive (aa) offspring
    recessive_recessive = n * (n - 1) * 1.0                          # aa x aa -> 100% chance of aa child
    hetero_hetero = m * (m - 1) * 0.25                               # Aa x Aa -> 25% chance of aa child
    hetero_recessive = (
        2 * (m * n) * 0.5
    )  # Aa x aa (or aa x Aa) -> 50% chance of aa child

    # Sum of recessive outcomes
    total_recessive_outcomes = (
        recessive_recessive + hetero_hetero + hetero_recessive
    )

    # Complementary probability: P(Dominant) = 1 - P(Recessive)
    p_dominant = 1 - (total_recessive_outcomes / total_pairs)
    return p_dominant


# Benchmark Dataset (Data from ROSALIND)
k_count, m_count, n_count = 29, 23, 26
result = get_dominant_probability(k_count, m_count, n_count)

print(f"{'Population Size (T):':<25} {k_count + m_count + n_count}")
print("-" * 33)
print(f"{'P(Dominant Phenotype):':<25} {result:.5f}")

Population Size (T):      78
---------------------------------
P(Dominant Phenotype):    0.77115


### Method 2: Dictionary-Mapped Parsing from Dataset File

In [99]:
# Read dataset file
with open("sample_data_probability.txt", "r") as file:
    k, m, n = map(int, file.read().split())

# Define genotype counts
genotypes = {
    "Homozygous Dominant (AA)": k,
    "Heterozygous (Aa)": m,
    "Homozygous Recessive (aa)": n,
}


total_pop = sum(genotypes.values())
total_pairs = total_pop * (total_pop - 1)


# Display genotype breakdown
for label, count in genotypes.items():
    print(f"{label:<26} : {count}")
print("-" * 31)
print(f"{'Total Population:':<26} : {total_pop}")


# Calculate probability of dominant phenotype
recessive_prob = (
    (n * (n - 1) * 1.0) + (m * (m - 1) * 0.25) + (2 * m * n * 0.5)
) / total_pairs
p_dominant = 1 - recessive_prob


print("-" * 31)
print(f"{'P(Dominant Phenotype):':<26} : {p_dominant:.5f}")

Homozygous Dominant (AA)   : 29
Heterozygous (Aa)          : 23
Homozygous Recessive (aa)  : 26
-------------------------------
Total Population:          : 78
-------------------------------
P(Dominant Phenotype):     : 0.77115


## Conclusion

- This project shows how to predict the exact chance that offspring will express a dominant trait in a randomly mating population using basic genetics and simple math. Instead of checking every possible parent pair, our Python program focuses only on the pairs that can produce a recessive ($aa$) child. This allows the script to calculate the exact probability instantly while using minimal memory.

- For the dataset of 78 individuals (29 AA, 23 Aa, 26 aa) from ROSALIND dataset, any randomly picked pair has a 77.115% chance (0.77115) of having an offspring with the dominant trait. This offers a fast, simple mathematical way to track traits, carrier rates, and gene patterns across generations.

- Key Results
Dominant Outcome: 
In a population of 78 individuals (29 AA, 23 Aa, 26 aa), any randomly chosen breeding pair has a 77.115% chance (0.77115) of producing an offspring with the dominant trait.

- Recessive Outcome: Only 22.885% (0.22885) of potential offspring will inherit two recessive alleles (aa) and show the recessive phenotype.

- Main Takeaway: 
Even when a large portion of the initial population is recessive (33.3%), the presence of dominant alleles (A) in AA and Aa parents ensures that most offspring will still express the dominant phenotype.